### PR TITLE
fix(compiler): bypass expansion form tokenizing in ngNonBindable blocks

### LINKED ISSUE #11859: ngNonBindable can't handle unescaped { characters
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior** 
`<div>Test binding: {{2 + 2}}</div>`
  => `Test binding: 4`
`<div ngNonBindable>Test non binding: {{2 + 2}}</div>`
  => `Test binding: {{2 + 2}}`
`<div>Test  non binding: {2 + 2}</div>`
  => Template parse errors, the "(Do you have an unescaped "{" in your template? Use "{{ '{' }}") to escape it.)" message

**Expected behavior**
The escaping of `{` characters isn't possible inside [ngNonBindable] elements so it's the compiler's job to ignore them altogether, I believe. 

`<div>Test  non binding: {2 + 2}</div>`
  => `Test binding: {2 + 2}`

**Reproduction of the problem**
here => http://plnkr.co/edit/GSzwEUyXyY1qDabjpEL1?p=preview

* **Angular version:** 2.0.0

### PR DESCRIPTION
**Pull request by tiagoroldao**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [not required] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
See #11859 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

fixes #11859

### CODE REVIEW

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 1]
```diff
@@ -88,6 +89,8 @@ class _Tokenizer {
   private _currentTokenType: TokenType;
   private _expansionCaseStack: TokenType[] = [];
   private _inInterpolation: boolean = false;
+  private _inNonBindableBlock: boolean = false;
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: Should this be a getter (depth gt 0)

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 2]
```diff
@@ -60,6 +60,7 @@ export function tokenize(
 }
 
 const _CR_OR_CRLF_REGEXP = /\r\n?/g;
+const NG_NON_BINDABLE_ATTR = 'ngNonBindable';
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: Could this be exported from the template parser ?

**Comment by tiagoroldao** [comment]
Reviewed comments and ammended, @vicb

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 3]
```diff
@@ -423,17 +426,26 @@ class _Tokenizer {
...
     let tagName: string;
     let lowercaseTagName: string;
+    let attributeStart: number;
     try {
       if (!chars.isAsciiLetter(this._peek)) {
         throw this._createError(_unexpectedCharacterErrorMsg(this._peek), this._getSpan());
       }
       const nameStart = this._index;
       this._consumeTagOpenStart(start);
+      if (this._inNonBindableBlock()) {
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: - merge this test with the code below,
- push all `AttrName` tokens to an array,

```ts
if (this._inNonBindableBlock()) {
  depth++;
} else if (attrNames.some(isNonBindableAttribute)) {
  depth = 1;
}
```

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 4]
```diff
@@ -640,6 +655,8 @@ class _Tokenizer {
     }
   }
 
+  private _inNonBindableBlock(): boolean { return this._nonBindableBlockDepth > 0; }
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: Block -> Element

**Comment by vicb** [comment]
Could you please make sure that we have integration test for 
- ICU like,
- interpolation like
in non bindable nodes ?

Thanks !

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 5]
```diff
@@ -523,6 +535,9 @@ class _Tokenizer {
     this._attemptCharCodeUntilFn(isNotWhitespace);
     this._requireCharCode(chars.$GT);
     this._endToken(prefixAndName);
+    if (this._inNonBindableBlock()) {
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: Probably not easy as the logic is in the parser...

May be we should move ICU logic into the parser ??

Comment (tiagoroldao) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: A first draft for clarity. Let me know if you prefer to move the logic/approach this in another way.
```javascript
private _checkNonBindableElementDepth () {
    let currentToken: Token = this.tokens[this.tokens.length - 1];

    if (currentToken.type === TokenType.TAG_OPEN_END || 
        currentToken.type === TokenType.TAG_OPEN_END_VOID || 
        currentToken.type === TokenType.TAG_CLOSE) {
      // check if this closes any previous, react
    }

    if (currentToken.type === TokenType.TAG_OPEN_END) {
      if (this._inNonBindableElement()) {
        this._nonBindableElementDepth++;
      }
      else if (/* tag contains ngNonBindable attr */){
        
      }
    }
    else if (currentToken.type === TokenType.TAG_CLOSE) {
      if (this._inNonBindableElement()) {
        this._nonBindableElementDepth--;
      }
    }
  };
```

**Comment by tiagoroldao** [comment]
@vicb changed the approach to keep (and cull) an tag token array, which ended up cleaner (code all in one place). None of the original code was kept, so please re-review as fresh.

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 6]
```diff
@@ -423,6 +426,7 @@ class _Tokenizer {
     let savedPos = this._savePosition();
     let tagName: string;
     let lowercaseTagName: string;
+    let attributeStart: number;
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: unused

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 7]
```diff
@@ -640,6 +647,67 @@ class _Tokenizer {
     }
   }
 
+  private _inNonBindableElement(): boolean { return this._nonBindableElementArr.length > 0; }
+
+  private _checkNonBindableElementDepth() {
+    let currentToken: Token = this.tokens[this.tokens.length - 1];
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: `_check...` makes me expect a boolean or a throw.
- find a better name
- add some inline docs `// ...`

Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: const currentToken

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 8]
```diff
@@ -640,6 +647,67 @@ class _Tokenizer {
...
+
+    if (this._inNonBindableElement() && (currentToken.type === TokenType.TAG_OPEN_END ||
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: `=== || === || ===` should probably be extracted in a fn

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 9]
```diff
@@ -711,3 +779,15 @@ function mergeTextTokens(srcTokens: Token[]): Token[] {
 
   return dstTokens;
 }
+
+function findPreviousToken(
+    array: Token[], filter: (token: Token) => boolean,
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: filter -> predicate ?

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 10]
```diff
@@ -711,3 +779,15 @@ function mergeTextTokens(srcTokens: Token[]): Token[] {
...
+    stopWhen: (token: Token) => boolean = t => false): Token {
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: lastPredicate

File: modules/@angular/compiler/src/ml_parser/lexer.ts [hunk 11]
```diff
@@ -711,3 +779,15 @@ function mergeTextTokens(srcTokens: Token[]): Token[] {
...
+  let index = array.length - 1;
+  while (index >= 0 && !stopWhen(array[index])) {
```
Comment (vicb) [inline modules/@angular/compiler/src/ml_parser/lexer.ts]: for ?

**Comment by vicb** [comment]
Added some comments but I'm not sure if this logic belongs here or in the parser. 

Have you tried / thought about detecting ICUs in the parser ?

**Comment by tiagoroldao** [comment]
Fixed the code according to your comments.

_Re: refactor_ Don't know if it should - moving tokenisation to the parser seems to further complicate things - wouldn't logic meant Interpolation detection also need to be moved?
Let me know what you think. If I go down that route will need to explore the parser further - may need time and help.

**Comment by tiagoroldao** [comment]
Also - The failed check seems unrelated to this work - is this correct?

**Comment by tiagoroldao** [comment]
@vicb, can you verify this PR? I disagree with the need for moving responsibilities, and assume the test failure was incidental and not related.

**Comment by tbosch** [comment]
We are closing this PR because it is unclear whether this is the right fix and we need to finalize the design before committing to a solution.

**Pull request closed without merging**